### PR TITLE
Design changes and tests for SQL DB replication commands

### PR DESCRIPTION
**Pull request by jaredmoo**

### CODE REVIEW

**Comment by jaredmoo** [comment]
```
(env) D:\git\azure-cli [replica ≡]> az sql db -h

Group
    az sql db: Manage databases.

Commands:
    copy               : Creates a copy of an existing database.
    create             : Creates a database.
    create-replica     : Creates a readable secondary replica database of an existing database.
    delete             : Deletes a database.
    delete-replica-link: Permanently stops data replication to a secondary replica.
    failover           : Fails over to a secondary replica. The secondary becomes the new primary.
    list               : Lists all databases in a server, or all databases in an elastic pool.
    list-replica-links : Lists a database's replication links.
    restore
    show               : Gets a database.
    update             : Updates a database.

(env) D:\git\azure-cli [replica ≡]> az sql db create-replica -h

Command
    az sql db create-replica: Creates a readable secondary replica database of an existing database.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server to create the new replica database in.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.
    --partner-elastic-pool        : Name of elastic pool to create the new replica database in.
    --partner-resource-group      : Name of the resource group to create the new replica database
                                    in. If unspecified, defaults to the origin resource group.
    --partner-service-objective   : Name of service objective for the new replica database.
    --tags                        : Resource tags.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db delete-replica-link -h

Command
    az sql db delete-replica-link: Permanently stops data replication to a secondary replica.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server that the other replica is in.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.
    --partner-resource-group      : Name of the resource group that the other replica is in. If
                                    unspecified, defaults to the first database's resource group.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db failover -h

Command
    az sql db failover: Fails over to a secondary replica. The secondary becomes the new primary.

Arguments
    --name -n           [Required]: Name of the database to fail over.
    --resource-group -g [Required]: Name of the resource group containing the secondary replica
                                    database that will become the new primary.
    --server -s         [Required]: Name of the server containing the secondary replica database
                                    that will become the new primary.
    --allow-data-loss             : If specified, the failover operation will allow data loss.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db list-replica-links -h

Command
    az sql db list-replica-links: Lists a database's replication links.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.
```

**Comment by devigned** [comment]
I think it would read better if you break out your commands into subgroups. Perhaps, something like the following command structure.
```
$ az sql db -h

Group
    az sql db: Manage databases.
Subgroups:
    replica
Commands:
    copy               : Creates a copy of an existing database.
    create             : Creates a database.
    delete             : Deletes a database.
    failover           : Fails over to a secondary replica. The secondary becomes the new primary.
    list               : Lists all databases in a server, or all databases in an elastic pool.
    restore
    show               : Gets a database.
    update             : Updates a database.
```

```
$ az sql db replica -h
Group
    az sql db replica: Manage database replicas.
Commands:
    create     : Creates a readable secondary replica database of an existing database.
    delete-link: Permanently stops data replication to a secondary replica.
    list-links : Lists a database's replication links.
```

```
$ az sql db replica link -h
Group
    az sql db replica link: Manage database replica links.
Commands:
    delete: Permanently stops data replication to a secondary replica.
    list : Lists a database's replication links.
```

```
$ az sql db replica link delete ...     # would break the replication link to the secondary replica
```

**Comment by jaredmoo** [comment]
Here's the conundrum: these commands are all related to replication, so `sql db replica` makes sense as a subgroup of related functionality. However a replica is itself a full database, there is no such ARM resource as a 'replica'. so arguably the noun should be just `sql db`. Both options make sense, I'm happy to hear arguments one way or the other.

I personally don't like putting `link` commands into their own subgroup because that feel like too much hierarchical depth - it forces me as a user to navigate with `-h` too many times to find what I want. If the CLI made it easier to view a command hierarchy this wouldn't bee to bad, but even then I think a subgroup with only 1 command in it feels like we overdid it. Your thoughts?

**Comment by jaredmoo** [comment]
Btw in later commit I changed `sql db failover` to `sql db set-primary-replica` to make it clearer how it works (you call in on a secondary replica to make it primary via failover mechanism).

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py [hunk 1]
```diff
@@ -208,7 +208,7 @@ def body(self):
         self.cmd('sql server firewall-rule list -g {} --server {}'
                  .format(rg, self.sql_server_name), checks=[JMESPathCheck('length(@)', 0)])
 
-        # test delete sql server
+        # test delete sql serverd
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py]: nit: typo here?

File: src/azure-cli-core/azure/cli/core/test_utils/vcr_test_base.py [hunk 2]
```diff
@@ -468,22 +468,43 @@ class ResourceGroupVCRTestBase(VCRTestBase):
...
+            self.resource_groups.append(_uniqueify_resource_group(resource_group + str(i + 2)))
+
         self.location = 'westus'
 
     def set_up(self):
-        self.cmd('group create --location {} --name {} --tags use=az-test'.format(
-            self.location, self.resource_group))
+        for g in self.resource_groups:
+            self.cmd('group create --location {} --name {} --tags use=az-test'.format(
+                self.location, g))
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/test_utils/vcr_test_base.py]: Yeah, this is useful. However, I'm going to send a PR for a new refurbished VCR test base which allows you to create arbitrary number of resource group and storage account.
Test code will end up looks like this: https://github.com/troydai/azure-cli/blob/venus/src/command_modules/azure-cli-acr/tests/test_acr_commands.py#L20

Comment (jaredmoo) [inline src/azure-cli-core/azure/cli/core/test_utils/vcr_test_base.py]: Sounds great. I'll take a look.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py (hunk 1 above)
Comment (jaredmoo) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py]: oops

**Comment by jaredmoo** [comment]
I tried `replica` subgroup as suggested by @devigned and massaged the help text so hopefully it's clear. I think I like it. Does this structure and help text make more sense?

```
(env) D:\git\azure-cli [replica ≡]> az sql db -h

Group
    az sql db: Manage databases.

Subgroups:
    replica: Manage replication between databases.

Commands:
    copy   : Creates a copy of an existing database.
    create : Creates a database.
    delete : Deletes a database.
    list   : Lists all databases in a server, or all databases in an elastic pool.
    restore
    show   : Gets a database.
    update : Updates a database.

(env) D:\git\azure-cli [replica ≡]> az sql db replica -h

Group
    az sql db replica: Manage replication between databases.

Commands:
    create     : Creates a database as a readable secondary replica of an existing database.
    delete-link: Permanently stops data replication between two database replicas.
    list-links : Lists the replicas of a database and corresponding replication status.
    set-primary: Sets which replica database is primary by failing over from the current primary
                 replica database.

(env) D:\git\azure-cli [replica ≡]> az sql db replica create -h

Command
    az sql db replica create: Creates a database as a readable secondary replica of an existing
    database.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server to create the new replica in.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.
    --partner-elastic-pool        : Name of elastic pool to create the new replica in.
    --partner-resource-group      : Name of the resource group to create the new replica in. If
                                    unspecified, defaults to the origin resource group.
    --partner-service-objective   : Name of service objective for the new replica.
    --tags                        : Resource tags.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db replica delete-link -h

Command
    az sql db replica delete-link: Permanently stops data replication between two database replicas.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --partner-server    [Required]: Name of the server that the other replica is in.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.
    --partner-resource-group      : Name of the resource group that the other replica is in. If
                                    unspecified, defaults to the first database's resource group.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db replica list-links -h

Command
    az sql db replica list-links: Lists the replicas of a database and corresponding replication
    status.

Arguments
    --name -n           [Required]: Name of the Azure SQL Database.
    --resource-group -g [Required]: Name of resource group.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

(env) D:\git\azure-cli [replica ≡]> az sql db replica set-primary -h

Command
    az sql db replica set-primary: Sets which replica database is primary by failing over from the
    current primary replica database.

Arguments
    --name -n           [Required]: Name of the database to fail over.
    --resource-group -g [Required]: Name of the resource group containing the secondary replica that
                                    will become the new primary.
    --server -s         [Required]: Name of the server containing the secondary replica that will
                                    become the new primary.
    --allow-data-loss             : If specified, the failover operation will allow data loss.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.
```

**Comment by jaredmoo** [comment]
You are right. I plan to do some design tweaks (including help) and functional tests for `restore` after https://github.com/Azure/azure-cli/pull/2351 is merged.

**Comment by jaredmoo** [comment]
Don't merge yet. I will do code review with replication experts on Wednesday to make sure the custom commands are correct. Thx

**Comment by jaredmoo** [comment]
I've reviewed with SQL replication team and it looks good. As soon as #2351 is merged I'll rebase this one and then we're ready to merge.

**Comment by jaredmoo** [comment]
Ready to merge

**Comment by troydai** [comment]
I'll take a second look and then merge.

File: src/azure-cli-core/azure/cli/core/test_utils/vcr_test_base.py (hunk 2 above)
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/test_utils/vcr_test_base.py]: Take a look at the example here: https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/test_acr_commands.py

**Comment by troydai** [comment]
Looks good to me. There are changes to the `vcr_test_base` could be risky because it will widely impact all the tests. But code the looks fine. We should keep an eye on how the tests behave in the next few days.

The issues addressed by the changes to the `vcr_test_base` are actually taken care of by the newly introduced `ScenarioTests` base test. Take a look at the ACR command module's tests.

The `vcr_test_base` will be eventually fade out in favor of the `ScenarioTests`. So let's avoid investing in it any further and make contributions to `ScenarioTests` instead.

**Comment by jaredmoo** [comment]
Thanks. Let me try rewriting this to use your new ScenarioTests test base. I was waiting for that to get merged and then forgot to try it out :)

**Comment by troydai** [comment]
@jaredmoo I haven't finish documentation yet, so it is doc free now. Any question shot me an mail.

**Comment by jaredmoo** [comment]
Assuming build validation passes, this is ready to merge. I rewrote all tests to use the new ScenarioTest base and it's working very nicely.

File: src/azure-cli-testsdk/azure/cli/testsdk/preparers.py [hunk 3]
```diff
@@ -154,7 +154,6 @@ def _get_resource_group(self, **kwargs):
             raise CliTestError(template.format(ResourceGroupPreparer.__name__,
                                                self.resource_group_parameter_name))
 
-
```
Comment (jaredmoo) [inline src/azure-cli-testsdk/azure/cli/testsdk/preparers.py]: oops. revert this

File: src/command_modules/azure-cli-sql/tests/test_sql_commands.py [hunk 4]
```diff
@@ -3,101 +3,128 @@
...
+                       'decorator @{} in front of this storage account preparer.'
+            raise CliTestError(template.format(ResourceGroupPreparer.__name__,
+                                               self.resource_group_parameter_name))
+
+
+class SqlServerMgmtScenarioTest(ScenarioTest):
+
+    @ResourceGroupPreparer()
+    def test_sql_server_mgmt(self, resource_group, resource_group_location):
+        servers = ['cliautomation51', 'cliautomation52']  # TODO: Server names should be randomized
```
Comment (troydai) [inline src/command_modules/azure-cli-sql/tests/test_sql_commands.py]: You can use `self.create_random_name`: https://github.com/Azure/azure-cli/blob/master/src/azure-cli-testsdk/azure/cli/testsdk/base.py#L85

Comment (troydai) [inline src/command_modules/azure-cli-sql/tests/test_sql_commands.py]: @jaredmoo I'll merge your PR now. You can follow up with the tests.

**Pull request merged**